### ISSUE TITLE
Ability to mark certain grain methods as having zero latency

### ISSUE THREAD

**Issue by JorgeCandeias** [OrleansContrib/OrleansDashboard#227]
When using the [reactive caching pattern](https://github.com/dotnet/orleans/tree/master/Samples/2.3/AdHocReactiveCaching), a grain's reactive poll style method can resolve the request *immediately* and *before* is has data to resolve it with. We do this, for example, by returning the `Task` from a `TaskCompletionSource`, which we set at a later time, when we do have data to fulfill it. This makes a *reactive poll* response have near-zero latency by design regardless of how long the requester of said method waits for a response or graceful refusal. *Reactive caching* is a behavioral contract and therefore the requester knows it may have to wait for up to the response timeout. This is also by design and it's fine.

However, the dashboard does not know about this and will therefore show abnormal latency metrics for said methods. On the positive side, Orleans itself does not complain about the reactive poll methods, as the underlying tasks take near-zero time on the scheduler and run on re-entrant grains.

This issue is here for adding an ability to the dashboard to mark these reactive poll methods as having zero latency in all requests and to gather views on how to best implement it. My current thinking, following suggestion from @jkonecki is to add a list of method names or infos to the dashboard options, to let the user decide what methods to always measure at zero latency.

Feel free to assign this to me for implementation.

**Comment by SebastianStehle** [comment]
Would it not be easier to customize the grain profiler? https://github.com/OrleansContrib/OrleansDashboard/blob/master/OrleansDashboard/Metrics/GrainProfiler.cs

**Comment by richorama** [comment]
Hey @JorgeCandeias, thanks for raising this.

I'm not sure if this is the right answer, but to support Orleankka was to add a hook to allow custom method names in the telemetry (all grain methods have the same name in Orleankka). I wonder if we could take a similar approach, perhaps to ignore certain method calls, or to give them a false value for latency.

Just an idea. I'll read up on the pattern, as I'm not familiar with what you're doing.

Feel free also to alter one of the test projects to demonstrate the problem.

**Comment by SebastianStehle** [comment]
I think what they are doing is the same as long polling. (https://www.pubnub.com/blog/2014-12-01-http-long-polling/)

**Comment by jkonecki** [comment]
@JorgeCandeias could you please explain what do you mean by 'abnormal latency'?

If the method to start long polling takes 20ms to execute and the actual result will be returned after 5sec than I would expect 20ms to be displayed in the Dashboard. This will allow me to notice any increase in the polling setup time, which I won't be able to do if I if ire that call or modify the latency to 0.

I haven't looked at your source code yet - do you use streams or grain observers?

**Comment by SebastianStehle** [comment]
My understanding: In long polling the connection can stay open for minutes, but resolving the actual value only takes a few ms. The equivalent of a connection is a method call in this case, which would also show several minutes in orleans dashboard.

e.g.

	public Task<int> GetValue()
	{
	   var tcs = new TaskCompletionSource<int>();
	   
	   this.currentTask = tcs;
	   
	   return tcs.Task;
	}

	public async Task OnTimerReceived()
	{
	   var value = await db.GetValue();
	   
	   this.currentTask?.SetCompleted(value);
	   this.currentTask = null;
	}

**Comment by jkonecki** [comment]
Wouldn't this block the grain (unless marked as reentrant)?

I would use grain observers (or streams) in this scenario.

**Comment by SebastianStehle** [comment]
Yes, this is what the post creator said:

> On the positive side, Orleans itself does not complain about the reactive poll methods, as the underlying tasks take near-zero time on the scheduler and run on re-entrant grains.

**Comment by SebastianStehle** [comment]
@JorgeCandeias Given your scenario, I think the best approach would be to have a delegate or attribute to prevent certain methods from being measured and then an implementation of the GrainProfiler that can be injected into your grain. This approach can also be used to measure internal metrics.

**Comment by jkonecki** [comment]
I'm assuming the grains were marked as reentrant for this very reason.

For me, grain observers would be a cleaner solution and wouldn't require
the whole grain to be reentrant. I tend to avoid reentrancy as much as
possible.

**Comment by SebastianStehle** [comment]
You are probably right, but I think that it would be good to have more control over the profiling.

I created a PR for that: https://github.com/OrleansContrib/OrleansDashboard/pull/228

See the example grain how to use it: https://github.com/OrleansContrib/OrleansDashboard/pull/228/files#diff-3a6be28c51a75a182088b16bee7d1069

**Comment by JorgeCandeias** [comment]
To answer a number of questions in one go:

*Reactive Caching for Composed Services* is a pattern developed at MSR with Orleans in mind.

* Here is the [Microsoft Research page](https://www.microsoft.com/en-us/research/publication/reactive-caching-for-composed-services/) on it.
* Here is the [white paper](https://www.microsoft.com/en-us/research/uploads/prod/2018/10/Burckhardt-Coppieters-OOPSLA2018-with-appendices.pdf).
* Here are the [conference slides](https://1drv.ms/f/s!AgoBH3oDy8d-iJ4zNvdBvFs-AvNgcA).
* Here is the [sample](https://github.com/dotnet/orleans/tree/master/Samples/2.3/AdHocReactiveCaching) in the Orleans repository.

TLDR; This stuff isn't coming out of the blue, people far smarter than me have come up with it. 😄 

Reactive caching caters for keeping two or more clients updated with the latest version of data *regardless of their relative network latency*. This works both within Orleans and between client machines and whatever front-end API you have. While faster clients may flicker faster, the slower client will catch up as best as network allows and adapt to any changing conditions. We use this pattern at my client to allow "streaming" of latest data snapshots both between Orleans grains and via REST to client machine browsers. It works fine.

> @richorama I wonder if we could take a similar approach, perhaps to ignore certain method calls, or to give them a false value for latency.

Ideally I would like to see those requests getting fulfilled in the dashboard - by the pattern nature, you can see massive spikes in the method chart whenever a reactive poll is resolved, as it resolves all pending requests at the same time - and that frequency data is of value. I just want the method to show zero actual latency. We have other ways of measuring actual completion latency on these special methods, e.g. via data snapshot timestamps - regardless, it's an application concern.

> @jkonecki could you please explain what do you mean by 'abnormal latency'?

A reactive poll method will *delay its response by design* until a new version of data is available, or send some graceful timeout response to the requester (.e.g null within Orleans or an *HTTP 204 No Content* via REST). Please check the [sample](https://github.com/dotnet/orleans/tree/master/Samples/2.3/AdHocReactiveCaching) to see how this works on the Orleans side. The dashboard does not know about this and just measures how long it took between the method starting and completing. In a very idle system, this latency will be the graceful timeout period you have configured. We therefore end up with a funny situation, where the Dashboard will show lower and lower average latency the more you pump the system with requests.

> @jkonecki I haven't looked at your source code yet - do you use streams or grain observers?

Neither. Reactive Caching is a request-response pattern where the target chooses to delay the response or not based on the input given. Please check the [sample](https://github.com/dotnet/orleans/tree/master/Samples/2.3/AdHocReactiveCaching). It's far simpler that it sounds.

> @SebastianStehle My understanding: In long polling the connection can stay open for minutes, but resolving the actual value only takes a few ms. The equivalent of a connection is a method call in this case, which would also show several minutes in orleans dashboard.

You got it. Except within Orleans the connections are already set up as part of the regular communication protocol. Orleans just simulates long polling via messages and callbacks. Your code shows the gist of it. The only additions in the pattern are that:

* The target grain will choose to delay the response or resolve it immediately based on input (e.g. the known data version on the requester side).
* The target grain will resolve the request regardless after some timeout (always under the Orleans timeout limit), with an agreed graceful timeout response, e.g. a null response value or some other default that makes sense to the client as meaning "got no updates yet, try again".
* The client grain keeps repeating the request until it itself deactivates.

> @jkonecki Wouldn't this block the grain (unless marked as reentrant)? I would use grain observers (or streams) in this scenario.

It would block if the grains were non-reentrant - reentrancy is a requirement for the pattern to work.

Grains observers, streams, or any other kind of push-only approach without the ability to conflate data on individual consumer back-pressure, cannot cater for significant differences in network latency on the consumer side. Even if you cater for back-pressure at the sender side, you end up catering to the slowest consumer, unless you develop some complicated code to cater for back-pressure from individual consumers. In addition to that, you also need to cater for subscriptions - who are you sending data to? are they alive? have heartbeats failed? When must we re-sub and un-sub them? With reactive caching, the request itself *is the heartbeat*, *is the subscription* and *is the back-pressure* - as long as the consumer plays along with the pattern and only issues the next long poll *after* dealing with its latest response as applicable. That's the trade-off by the way, it does require a proactive consumer.

> @SebastianStehle Given your scenario, I think the best approach would be to have a delegate or attribute to prevent certain methods from being measured and then an implementation of the GrainProfiler that can be injected into your grain. This approach can also be used to measure internal metrics.

That does sound nicer than manually adding method infos to an options class. However, wouldn't that force a Dashboard dependency on the user's grain interfaces project? Not that I mind.

> @jkonecki For me, grain observers would be a cleaner solution and wouldn't require
the whole grain to be reentrant. I tend to avoid reentrancy as much as
possible.

As explained above. It's a design trade-off.

**Comment by JorgeCandeias** [comment]
> You are probably right, but I think that it would be good to have more control over the profiling.
> 
> I created a PR for that: #228
> 
> See the example grain how to use it: https://github.com/OrleansContrib/OrleansDashboard/pull/228/files#diff-3a6be28c51a75a182088b16bee7d1069

Nice. I'll check this out as soon as I can.

**Comment by SebastianStehle** [comment]
Can this be closed now, the PR is merged

**Comment by JorgeCandeias** [comment]
@SebastianStehle Just had the chance to test this, works like a charm, thank you for the PR. 👍
